<a href="https://colab.research.google.com/github/thakurakanksha288/AI-ASSIGNMENT-PYTHON-/blob/main/Data_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
path="/content/students_performance_raw.csv"
df=pd.read_csv(path)
df.head()

student_id          name    age  gender  maths_score  science_score  \
0           1  Rahul Sharma   21.0    Male         88.0             90   
1           2        Sunita    NaN       F          NaN             85   
2           3           NaN   22.0    male         75.0             -1   
3           4       rohan k  300.0  Female         60.0             58   
4           5           NaN   20.0  female         55.0             62   

   english_score attendance placement_interest  
0           92.0        92%                Yes  
1           80.0        85%                  Y  
2           70.0        NaN              Maybe  
3           65.0        70%                 No  
4            NaN        95%                yes

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [ ]:
import numpy as np
import pandas as pd

# Load the dataset from the specified path
path="/content/students_performance_raw.csv"
df=pd.read_csv(path)

# Display the first few rows of the raw DataFrame
df.head()

# Calculate the median age to use for filling missing age values
median=df['age'].median()
print(f"Median age: {median}")

# --- Data Cleaning and Preprocessing Steps ---

# Fill NaN values in the 'age' column with the calculated median
df['age'] = df['age'].fillna(median)

# Handle missing and invalid values in score columns ('maths_score', 'science_score', 'english_score')
score_columns = ['maths_score', 'science_score', 'english_score']
for col in score_columns:
    # Replace '-1' entries with NaN, as -1 is likely an indicator of missing data
    df[col] = df[col].replace(-1, np.nan)
    # Fill any remaining NaN values (original NaNs and those from -1) with the mean of the column
    df[col] = df[col].fillna(df[col].mean())
    # Enforce score constraints: not negative and not above 100
    df[col] = df[col].clip(lower=0, upper=100)

# Standardize the 'gender' column:
# 1. Replace 'F' with 'female' and 'Male' with 'male'
# 2. Convert all entries to lowercase for consistency
df['gender'] = df['gender'].replace({'F': 'female', 'Male': 'male'}).str.lower()

# Fill NaN values in the 'name' column with 'Unknow'
df['name'] = df['name'].fillna('Unknow')

# Standardize the 'placement_interest' column:
# 1. Replace 'YES', 'Y' with 'yes'
# 2. Replace 'NO', 'N', 'Maybe' with 'no'
# 3. Convert all entries to lowercase for consistency
df['placement_interest'] = df['placement_interest'].replace({'YES': 'yes', 'NO': 'no','Maybe':'no','Y':'yes','N':'no'}).str.lower()

# Correct age values that are 100 or greater by replacing them with the median age
# This assumes ages 100+ are data entry errors
df.loc[df['age'] >= 100, 'age'] = df['age'].median()

# Handle the 'attendance' column:
# 1. Remove the '%' character
# 2. Convert the column to a float data type
# 3. Calculate the mean attendance from the cleaned numeric values
# 4. Fill any NaN values with the calculated mean attendance
# 5. Round the attendance values and convert to integer type
df['attendance'] = df['attendance'].str.replace('%', '', regex=False).astype(float)
mean_attendance = df['attendance'].mean()
df['attendance'] = df['attendance'].fillna(mean_attendance)
df['attendance'] = df['attendance'].round().astype(int)

# Display the first few rows of the cleaned DataFrame
df.head()

Median age: 21.5


student_id          name   age  gender  maths_score  science_score  \
0           1  Rahul Sharma  21.0    male    88.000000           90.0   
1           2        Sunita  21.5  female    81.333333           85.0   
2           3        Unknow  22.0    male    75.000000           82.0   
3           4       rohan k  21.5  female    60.000000           58.0   
4           5        Unknow  20.0  female    55.000000           62.0   

   english_score  attendance placement_interest  
0           92.0          92                yes  
1           80.0          85                yes  
2           70.0          89                 no  
3           65.0          70                 no  
4           85.0          95                yes

# Task
Import the pandas library, load the `house_price_raw(1).csv` dataset into a DataFrame, and then display its first few rows, along with its information summary and descriptive statistics.

## Inspect House Price Data

### Subtask:
Examine the `house_price_raw(1).csv` dataset by displaying its head, info, and descriptive statistics to identify potential issues like missing values, incorrect data types, or outliers.


**Reasoning**:
The previous attempt to load the data failed because pandas was not imported. This step will re-import pandas and then load the dataset, display its head, info, and descriptive statistics as requested in the instructions to inspect the data.



In [2]:
import pandas as pd

path="/content/house_price_raw(1).csv"
df=pd.read_csv(path)

print("First 5 rows of the DataFrame:")
display(df.head())

print("\nDataFrame Info:")
df.info()

print("\nDescriptive Statistics:")
display(df.describe())

First 5 rows of the DataFrame:


size_sqft  bedrooms   age   price  location       furnishing
0     1200.0         2  10.0  100000     Urban  Fully-furnished
1     1500.0         3   5.0  150000     Urban   Semi-furnished
2        NaN         1  15.0   80000     Rural              NaN
3      800.0         1   NaN   60000     Rural      Unfurnished
4     2000.0         4  30.0  350000  Suburban   Semi-furnished


DataFrame Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   size_sqft   6 non-null      float64
 1   bedrooms    8 non-null      int64  
 2   age         7 non-null      float64
 3   price       8 non-null      int64  
 4   location    8 non-null      object 
 5   furnishing  6 non-null      object 
dtypes: float64(2), int64(2), object(2)
memory usage: 516.0+ bytes

Descriptive Statistics:


size_sqft  bedrooms        age          price
count     6.000000  8.000000   7.000000       8.000000
mean   1375.000000  2.250000  14.285714  143750.000000
std     475.131561  1.035098   8.459990   94254.594431
min     800.000000  1.000000   5.000000   60000.000000
25%    1012.500000  1.750000   9.000000   87500.000000
50%    1350.000000  2.000000  12.000000  110000.000000
75%    1725.000000  3.000000  17.500000  162500.000000
max    2000.000000  4.000000  30.000000  350000.000000

## Clean House Price Data

### Subtask:
Based on the inspection, apply appropriate data cleaning techniques to the house price dataset. This may include handling missing values (e.g., imputation or removal), correcting data types, and addressing any inconsistencies or outliers found.


**Reasoning**:
To handle missing values as per the instructions, I will first calculate the median for 'size_sqft' and 'age' columns and the mode for the 'furnishing' column. Then, I will use these calculated values to fill the respective missing (NaN) entries in the DataFrame. Finally, I'll display the head and info of the cleaned DataFrame to verify the changes.



In [4]:
import numpy as np

# 1. Calculate the median for 'size_sqft' and fill NaN values
median_size_sqft = df['size_sqft'].median()
df['size_sqft'] = df['size_sqft'].fillna(median_size_sqft)

# 2. Calculate the median for 'age' and fill NaN values
median_age = df['age'].median()
df['age'] = df['age'].fillna(median_age)

# 3. Calculate the mode for 'furnishing' and fill NaN values
# .mode()[0] is used because mode() can return multiple values if there's a tie
mode_furnishing = df['furnishing'].mode()[0]
df['furnishing'] = df['furnishing'].fillna(mode_furnishing)

print("DataFrame after filling missing values:")
display(df.head())

print("\nDataFrame Info after cleaning:")
df.info()

DataFrame after filling missing values:


size_sqft  bedrooms   age   price  location       furnishing
0     1200.0         2  10.0  100000     Urban  Fully-furnished
1     1500.0         3   5.0  150000     Urban   Semi-furnished
2     1350.0         1  15.0   80000     Rural  Fully-furnished
3      800.0         1  12.0   60000     Rural      Unfurnished
4     2000.0         4  30.0  350000  Suburban   Semi-furnished


DataFrame Info after cleaning:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8 entries, 0 to 7
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   size_sqft   8 non-null      float64
 1   bedrooms    8 non-null      int64  
 2   age         8 non-null      float64
 3   price       8 non-null      int64  
 4   location    8 non-null      object 
 5   furnishing  8 non-null      object 
dtypes: float64(2), int64(2), object(2)
memory usage: 516.0+ bytes


## Display Cleaned Data

### Subtask:
Show the first few rows of the cleaned house price DataFrame to confirm that the data cleaning process was successful and the data is now in a suitable format.


## Final Task

### Subtask:
Summarize the cleaning steps performed on the `house_price_raw(1).csv` dataset and present the cleaned data.


## Summary:

### Q&A
The cleaning steps performed on the `house_price_raw(1).csv` dataset involved handling missing values in three columns:
1.  **`size_sqft`**: Missing values were imputed using the median of the column.
2.  **`age`**: Missing values were imputed using the median of the column.
3.  **`furnishing`**: Missing values were imputed using the mode of the column.

After these steps, all specified missing values were addressed, and the dataset was confirmed to have no missing entries in the `size_sqft`, `age`, and `furnishing` columns. The cleaned data's first few rows were displayed to confirm the successful cleaning.

### Data Analysis Key Findings
*   The initial dataset contained 8 entries and 6 columns.
*   Missing values were identified in three columns: `size_sqft` (2 missing), `age` (1 missing), and `furnishing` (2 missing).
*   Data types were `float64` for `size_sqft` and `age`, `int64` for `bedrooms` and `price`, and `object` for `location` and `furnishing`.
*   Numerical columns showed varying ranges: `size_sqft` from 800 to 2000 sqft (average 1375 sqft), `bedrooms` from 1 to 4 (average 2.25), `age` from 5 to 30 years (average \~14.29 years), and `price` from \$60,000 to \$350,000 (average \$143,750).
*   All missing values were successfully imputed: `size_sqft` and `age` were filled with their respective medians, and `furnishing` was filled with its mode.
*   Post-cleaning, the dataset no longer contained any missing values across all columns, with each column showing 8 non-null entries.

### Insights or Next Steps
*   The data is now clean and ready for further analysis, such as exploratory data analysis, feature engineering, or predictive modeling.
*   Consider investigating the impact of the imputed values on the overall data distribution and potential biases introduced, especially if the proportion of missing data were higher.
